optuna/progress_bar.py

```diff
@@ -0,0 +1,76 @@
+import logging
+from typing import Any
+from typing import Optional
+
+from tqdm.auto import tqdm
+
+from optuna import logging as optuna_logging
+
+_tqdm_handler = None  # type: Optional[_TqdmLoggingHandler]
+
+
+# Reference: https://gist.github.com/hvy/8b80c2cedf02b15c24f85d1fa17ebe02
+class _TqdmLoggingHandler(logging.StreamHandler):
+
+    def emit(self, record: Any) -> None:
+        try:
+            msg = self.format(record)
+            tqdm.write(msg)
+            self.flush()
+        except (KeyboardInterrupt, SystemExit):
+            raise
+        except Exception:
+            self.handleError(record)
+
+
+class _ProgressBar:
```

Comment:
Reverting my previous suggestion, sorry. I'm wondering by the way if we shouldn't ignore this hacking warning. It's a part of their Python 2 --> 3 migration support (https://github.com/openstack/hacking/blob/f35a26f44beeb5c27f55f8c2e965dcffc10e8287/HACKING.rst) which shouldn't be non-relevant to us not that Python 2 is already dropped.
```suggestion
class _ProgressBar(object):
```

Reply:
another reference is https://www.python.org/doc/newstyle/.